### PR TITLE
Updates Java example

### PR DESCRIPTION
**Pull request by seancook**
* Uses gradle for building and dependencies
* Self-contained example, runnable via `./gradlew appRun`
* Adds tests, runnable via `./gradlew test`

### CODE REVIEW

**Comment by seancook** [comment]
@blakewatters @dwolf07 @maju6406 - please review when you have a moment :family:

**Comment by blakewatters** [comment]
.DS_Store files should be in the .gitignore and stripped from repo

File: identity-services-samples/java/layer/src/main/java/com/layer/LayerConfig.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Your Layer app's configuration values.
+ *
+ * All values are available in the Layer Dashboard for your project.
+ */
+public class LayerConfig {
+    private final static String LAYER_KEY_ID =
+            "layer:///keys/e3b73796-3a1b-11e5-b8be-fdebe7056b10";
+    private final static String LAYER_PROVIDER_ID =
+            "layer:///providers/251c5b9c-cc41-11e4-9872-eae8a6002c37";
+    private final static String LAYER_RSA_KEY_PATH =
```
Comment (mubarak) [inline identity-services-samples/java/layer/src/main/java/com/layer/LayerConfig.java]: i think this constant value is incorrect (i think `.pk8` should come from external property file?)

File: identity-services-samples/java/layer/src/main/java/com/layer/LayerIdentityTokenGenerator.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     * @return a byte array of the key's contents
+     * @throws FileNotFoundException if the given path is not found
+     * @throws IOException           if an error occurs reading from the file
+     */
+    private byte[] readPrivateKeyFromDisk(final String path) throws IOException {
+        final File privateKeyFile = new File(path);
+        final DataInputStream dis = new DataInputStream(new FileInputStream(privateKeyFile));
+        final byte[] privateBytes = new byte[(int) privateKeyFile.length()];
+        dis.readFully(privateBytes);
+        dis.close();
```
Comment (mubarak) [inline identity-services-samples/java/layer/src/main/java/com/layer/LayerIdentityTokenGenerator.java]: i think it is better to wrap `readFully` inside `try` and call `dis.close()` inside `finally` block to release system resources associated with the stream (if `readFully` throws an exception `dis` won't be closed).

File: identity-services-samples/java/web/src/main/java/IdentityTokenServlet.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+     * @param token    the identity token to return
+     * @throws IOException if we encounter a problem writing our response
+     */
+    private void returnIdentityToken(HttpServletResponse response, String token)
+            throws IOException {
+
+        final JsonObject outJson = new JsonObject();
+        outJson.addProperty("identity_token", token);
+        response.setContentType("application/json");
+        response.setCharacterEncoding("UTF8");
```
Comment (mubarak) [inline identity-services-samples/java/web/src/main/java/IdentityTokenServlet.java]: `UTF-8`?

File: identity-services-samples/java/layer/src/main/java/com/layer/LayerConfig.java (hunk 1 above)
Comment (seancook) [inline identity-services-samples/java/layer/src/main/java/com/layer/LayerConfig.java]: thanks, these were constants i was using for testing and should've removed. each of these values needs to be completed by the developer, so setting them to empty strings.

File: identity-services-samples/java/web/src/main/java/IdentityTokenServlet.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    private JsonObject readJsonFromRequest(HttpServletRequest request)
+            throws IllegalStateException {
+
+        final StringBuilder stringBuilder = new StringBuilder();
+        String line;
+        try {
+            BufferedReader reader = request.getReader();
+            while ((line = reader.readLine()) != null) {
+                stringBuilder.append(line);
+            }
```
Comment (mubarak) [inline identity-services-samples/java/web/src/main/java/IdentityTokenServlet.java]: `reader.close()`?

if you use Java 7 then `try_with_resource`:  https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Comment (seancook) [inline identity-services-samples/java/web/src/main/java/IdentityTokenServlet.java]: thanks @mubarak! i didn't know about try_with_resource; will make code much cleaner

File: identity-services-samples/java/README.md [hunk 5]
```diff
@@ -1,28 +1,77 @@
...
 
-* `PROVIDER_ID` - Provider ID found in the Layer Dashboard under "Authentication"
-* `KEY_ID` - Public key generated and stored in the Layer Dashboard under "Authentication"
-* `RSA_KEY_LOCATION` - Path to the PK8 file generated from the RSA key
+Example successful response:
 
-### Parameters:
+```JSON
+{
+  'identity_token': 'eyJ0eXAiOiJKV1Mi...'
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: Double quotes.  Github red-ifies this.

File: identity-services-samples/java/README.md [hunk 6]
```diff
@@ -1,28 +1,77 @@
+# Overview
 
-The code in this folder represents a simple backend to generate identity tokens using a Java Servlet.
+The code in this folder provides an example for creating a Layer Identity Token using an HttpServlet.
 
-### Setup:
+The example endpoint, `identity-token`, that requires the following parameters:
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: nix `that`

File: identity-services-samples/java/README.md [hunk 7]
```diff
@@ -1,28 +1,77 @@
...
-{
-"identity_token": "eyJ0eXAiOiJKV1MiLCJhbGciOiJSUzI1NiIsImN0eSI6ImxheWVyLWVpdDt2PTEiLCJraWQiOiI2OWZkZDVhYS02NDc4LTExZTQtOTdmMS0xZGVkMDAwMDAwZTYifQ.eyJpc3MiOiI1YTczMWE0Yy02M2JlLTExZTQtOTEyNC1hYWE1MDIwMDc1ZjgiLCJpYXQiOjE0MTUxNTExMTcsImV4cCI6MTQxNTE2MTExNywibmNlIjoibkcrWHZFb0c3dDBXSEZrZjN0QmlWdEdjekFsTXArUmwydXVqWkN0TVJsSEcxb1FVU05BSXZnM0ZTbnFzTDhiNlFFK2pIZU8vZHJsZ2FJNXJnRXVIR2c9PSJ9.SPVPzzN7S09OafQZV7E_LvHF1mmvj5VU0Kn780ef4tegwLUS_pYj7ODfgSZPS2-MNRFhYb5ACZqtoxNkv32CJBzJzFuwBkZ3CsuX8xdpeXWEqvYtK2OV73x1TNA8RGmWyVjVKq7xjpGUORkLk7KQW3QRrQGSVx_jeOiUxb9HvZI"
-}
-```
\ No newline at end of file
+### Configure your Layer app
+
+You will then need to define the following variables in the `LayerConfig` class:
+
+* `LAYER_PROVIDER_ID` - Provider ID found in the Layer Dashboard under "Authentication"
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: There is no "Authentication" link on the dashboard.  I'd rephrase:
Provider ID found in the Layer Dashboard under "Keys" -> "Account: Provider ID"

File: identity-services-samples/java/README.md [hunk 8]
```diff
@@ -1,28 +1,77 @@
...
+* `LAYER_KEY_ID` - Public key generated and stored in the Layer Dashboard under "Authentication"
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: Same deal here, but this one makes it sound like you should use the actual public key.  I'd rephrase (this may need some reduction):
ID of the public key generated and stored in the Layer Dashboard under "Keys" -> "Account: Authentication keys"

File: identity-services-samples/java/README.md [hunk 9]
```diff
@@ -1,28 +1,77 @@
...
+openssl pkcs8 -topk8 -nocrypt -outform DER -in layer.pem -out layer.pk8```
+
+
+## Running the servlet
+
+To run the servlet locally, enter the following command:
+
+```console
+./gradlew appRun
+```
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: This markdown doesn't render properly.

File: identity-services-samples/java/README.md [hunk 10]
```diff
@@ -1,28 +1,77 @@
...
+You will then need to define the following variables in the `LayerConfig` class:
+
+* `LAYER_PROVIDER_ID` - Provider ID found in the Layer Dashboard under "Authentication"
+* `LAYER_KEY_ID` - Public key generated and stored in the Layer Dashboard under "Authentication"
+* `LAYER_RSA_KEY_PATH` - Path to the PK8 version of your RSA key (see below)
+
+__Note:__ You must convert your private key to PKCS8 format so that Java can read it:
+
+```console
+openssl pkcs8 -topk8 -nocrypt -outform DER -in layer.pem -out layer.pk8```
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: Missing closing quotes

File: identity-services-samples/java/README.md [hunk 11]
```diff
@@ -1,28 +1,77 @@
...
+Content-Type: application/json;charset=UTF8
+Transfer-Encoding: chunked
+Date: Wed, 12 Aug 2015 03:47:29 GMT
+
+{"identity_token":"...SNIP..."}
+```
+
+### Verify the generated token
+
+You should verify the output of the signing request by visiting the **Tools** section of the [Layer dashboard](https://developer.layer.com/dashboard/). Paste the value of the `identity_token` key you received from the output above and click `validate`. You should see "Token valid."
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: Capitalize "Dashboard" for consistency with mentions above.

File: identity-services-samples/java/README.md [hunk 12]
```diff
@@ -1,28 +1,77 @@
...
 
-```json
-{
-"identity_token": "eyJ0eXAiOiJKV1MiLCJhbGciOiJSUzI1NiIsImN0eSI6ImxheWVyLWVpdDt2PTEiLCJraWQiOiI2OWZkZDVhYS02NDc4LTExZTQtOTdmMS0xZGVkMDAwMDAwZTYifQ.eyJpc3MiOiI1YTczMWE0Yy02M2JlLTExZTQtOTEyNC1hYWE1MDIwMDc1ZjgiLCJpYXQiOjE0MTUxNTExMTcsImV4cCI6MTQxNTE2MTExNywibmNlIjoibkcrWHZFb0c3dDBXSEZrZjN0QmlWdEdjekFsTXArUmwydXVqWkN0TVJsSEcxb1FVU05BSXZnM0ZTbnFzTDhiNlFFK2pIZU8vZHJsZ2FJNXJnRXVIR2c9PSJ9.SPVPzzN7S09OafQZV7E_LvHF1mmvj5VU0Kn780ef4tegwLUS_pYj7ODfgSZPS2-MNRFhYb5ACZqtoxNkv32CJBzJzFuwBkZ3CsuX8xdpeXWEqvYtK2OV73x1TNA8RGmWyVjVKq7xjpGUORkLk7KQW3QRrQGSVx_jeOiUxb9HvZI"
-}
-```
\ No newline at end of file
+### Configure your Layer app
+
+You will then need to define the following variables in the `LayerConfig` class:
```
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: s/LayerConfig/com.layer.LayerConfig/   (not that there's much confusion, but hey)

File: identity-services-samples/java/README.md (hunk 11 above)
Comment (dcheckoway) [inline identity-services-samples/java/README.md]: Remove the word "key" -- actually I'd simplify that sentence:

Paste the `identity_token` value into the form on that page and click `Validate`.

(note I capitalized V to match the UI)

File: identity-services-samples/java/layer/src/main/java/com/layer/LayerIdentityTokenGenerator.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        final File privateKeyFile = new File(path);
+        final FileInputStream fileInputStream = new FileInputStream(privateKeyFile);
+        final DataInputStream dis = new DataInputStream(fileInputStream);
+        final byte[] privateBytes = new byte[(int) privateKeyFile.length()];
+        try {
+            dis.readFully(privateBytes);
+        } catch (IOException ioe) {
+            /** No-op **/
+        } finally {
+            dis.close();
```
Comment (dcheckoway) [inline identity-services-samples/java/layer/src/main/java/com/layer/LayerIdentityTokenGenerator.java]: It's really `fileInputStream` that needs to be closed.

Comment (seancook) [inline identity-services-samples/java/layer/src/main/java/com/layer/LayerIdentityTokenGenerator.java]: Thanks!
